pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,493 @@
+func (m *NodeLatencyMonitor) onNodeLatencyMonitorAdd(obj interface{}) {
+	nlm := obj.(*v1alpha1.NodeLatencyMonitor)
+	klog.InfoS("NodeLatencyMonitor added", "NodeLatencyMonitor", klog.KObj(nlm))
+
+	if err := m.updateLatencyConfig(nlm); err != nil {
+		klog.ErrorS(err, "Failed to update latency config")
+	}
+}
+
+// onNodeLatencyMonitorUpdate is the event handler for updating NodeLatencyMonitor.
+func (m *NodeLatencyMonitor) onNodeLatencyMonitorUpdate(oldObj, newObj interface{}) {
+	oldNLM := oldObj.(*v1alpha1.NodeLatencyMonitor)
+	newNLM := newObj.(*v1alpha1.NodeLatencyMonitor)
+	klog.InfoS("NodeLatencyMonitor updated", "NodeLatencyMonitor", klog.KObj(newNLM))
+
+	if oldNLM.GetGeneration() == newNLM.GetGeneration() {
+		return
+	}
+
+	if err := m.updateLatencyConfig(newNLM); err != nil {
+		klog.ErrorS(err, "Failed to update latency config")
+	}
+}
+
+func (m *NodeLatencyMonitor) updateLatencyConfig(nlm *v1alpha1.NodeLatencyMonitor) error {
+	// Parse the ping interval
+	pingInterval := time.Duration(nlm.Spec.PingIntervalSeconds) * time.Second
+
+	// Update the latency config
+	m.latencyConfig = &LatencyConfig{
+		Enable:   true,
+		Interval: pingInterval,
+	}
+
+	// Notify the latency config changed
+	m.latencyConfigChanged <- struct{}{}
+
+	return nil
+}
+
+// onNodeLatencyMonitorDelete is the event handler for deleting NodeLatencyMonitor.
+func (m *NodeLatencyMonitor) onNodeLatencyMonitorDelete(obj interface{}) {
+	// Update the latency config
+	m.latencyConfig = &LatencyConfig{Enable: false}
+
+	// Notify the latency config changed
+	m.latencyConfigChanged <- struct{}{}
+}
+
+func (m *NodeLatencyMonitor) sendPing(socket net.PacketConn, addr net.IP) error {
+	var requestType icmp.Type
+
+	// Resolve the IP address
+	ip := &net.IPAddr{IP: addr}
+
+	// Create a new ICMP packet
+	if addr.To4() == nil {
+		requestType = ipv6.ICMPTypeEchoRequest
+	} else {
+		requestType = ipv4.ICMPTypeEcho
+	}
+
+	timeStart := time.Now()
+	seqID := getICMPSeq()
+	body := &icmp.Echo{
+		ID:   os.Getpid() & 0xffff,                       // Use the current process ID as the ICMP ID
+		Seq:  int(seqID),                                 // Use the current sequence number as the ICMP sequence number
+		Data: []byte(timeStart.Format(time.RFC3339Nano)), // Store the current time in the ICMP data
+	}
+	msg := icmp.Message{
+		Type: requestType,
+		Code: 0,
+		Body: body,
+	}
+	klog.InfoS("Send ICMP message", "IP", ip, "SeqID", seqID, "body", body)
+
+	// Serialize the ICMP message
+	msgBytes, err := msg.Marshal(nil)
+	if err != nil {
+		return err
+	}
+
+	// Send the ICMP message
+	_, err = socket.WriteTo(msgBytes, ip)
+	if err != nil {
+		return err
+	}
+
+	// Store current send info
+	// Read the ICMP message
+	oldEntry := m.latencyStore.GetNodeIPLatencyEntryByKey(addr.String())
+	if oldEntry != nil {
+		// Update the latency store
+		oldEntry.LastSendTime = timeStart
+		oldEntry.SeqID = seqID
+		m.latencyStore.UpdateNodeIPLatencyEntryByKey(addr.String(), oldEntry)
```

Comment:
I don't understand this code: `GetNodeIPLatencyEntryByKey` is returning a pointer to the `NodeIPLatencyEntry` value. We are then modifying the value directly through the pointer , before calling `UpdateNodeIPLatencyEntryByKey`. But by modifying the pointer we already have mutated the entry, so there is no need to call `UpdateNodeIPLatencyEntryByKey` after that.

Either the operation of mutating the entry through the pointer is thread-safe (i.e., you can prove that there is no concurrent access to the entry through another goroutine), in which case the call to `UpdateNodeIPLatencyEntryByKey` is useless, or the mutation is not thread-safe, in which case `GetNodeIPLatencyEntryByKey` should return a value and not a pointer (but there can be other implications if you do that).

One generic way to handle what you need would be to have a function that updates (or creates if needed) the entry using an arbitrary mutator function:
```go
func (l *LatencyStore) SetNodeIPLatencyEntry(key string, mutator func(entry *NodeIPLatencyEntry)) {
	l.mutex.Lock()
	defer l.mutex.Unlock()

	entry, ok := l.nodeIPLatencyMap[key]
        if !ok {
		entry = &NodeIPLatencyEntry{}
		l.nodeIPLatencyMap[key] = entry
        }
        mutator(entry)
} 
```

Reply:
Thanks, this writeup does go a long way to avoiding competition issues, I've updated it.